src/sink/gpub.rs

```diff
@@ -0,0 +1,264 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # File Offramp
+//!
+//! Writes events to a file, one event per line
+//!
+//! ## Configuration
+//!
+//! See [Config](struct.Config.html) for details.
+
+#![cfg(not(tarpaulin_include))]
+
+use crate::connectors::gcp::{pubsub, pubsub_auth};
+use crate::connectors::qos::{self, QoSFacilities, SinkQoS};
+use crate::sink::prelude::*;
+use futures::executor::block_on;
+use googapis::google::pubsub::v1::publisher_client::PublisherClient;
+use halfbrown::HashMap;
+use tonic::transport::Channel;
+use tremor_pipeline::{EventIdGenerator, OpMeta};
+use tremor_value::Value;
+
+pub struct GoogleCloudPubSub {
+    #[allow(dead_code)]
+    config: Config,
+    remote: Option<PublisherClient<Channel>>,
+    is_down: bool,
+    qos_facility: Box<dyn SinkQoS>,
+    reply_channel: Option<Sender<sink::Reply>>,
+    is_linked: bool,
+    postprocessors: Postprocessors,
+    codec: Box<dyn Codec>,
+    event_id_gen: EventIdGenerator,
+}
+
+#[derive(Deserialize)]
+pub struct Config {}
+
+enum PubSubCommand {
+    SendMessage(String, String, Value<'static>),
+    Unknown,
+}
+
+impl std::fmt::Display for PubSubCommand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match *self {
+                PubSubCommand::SendMessage(_, _, _) => "send_message",
+                PubSubCommand::Unknown => "Unknown",
+            }
+        )
+    }
+}
+
+impl ConfigImpl for Config {}
+
+impl offramp::Impl for GoogleCloudPubSub {
+    fn from_config(config: &Option<OpConfig>) -> Result<Box<dyn Offramp>> {
+        if let Some(config) = config {
+            let config: Config = Config::new(config)?;
+            let remote = Some(block_on(pubsub_auth::setup_publisher_client())?);
+            let hostport = "pubsub.googleapis.com:443";
+            Ok(SinkManager::new_box(Self {
+                config,
+                remote,
+                is_down: false,
+                qos_facility: Box::new(QoSFacilities::recoverable(hostport.to_string())),
+                reply_channel: None,
+                is_linked: false,
+                postprocessors: vec![],
+                codec: Box::new(crate::codec::null::Null {}),
+                event_id_gen: EventIdGenerator::new(0), // Fake ID overwritten in init
+            }))
+        } else {
+            Err("Offramp Google Cloud Pubsub (pub) requires a config".into())
+        }
+    }
+}
+
+macro_rules! parse_arg {
+    ($field_name: expr, $o: expr) => {
+        if let Some(Value::String(snot)) = $o.get($field_name) {
+            snot.to_string()
+        } else {
+            return Err(format!("Invalid Command, expected `{}` field", $field_name).into());
+        }
+    };
+}
+
+fn parse_command(value: &Value) -> Result<PubSubCommand> {
+    if let Value::Object(o) = value {
```

Comment:
we can simplifyt this with https://docs.rs/value-trait/0.2.4/value_trait/trait.ValueAccess.html#tymethod.as_object

```
let o = value.as_object().ok_or("Invalid Command")?;
```